### PR TITLE
Feature/wa 271 refactoring authentication service

### PR DESCRIPTION
**Pull request by DmitryBespalov**
Major changes:

- `Account` is removed 🔥 
- `Clock` and `Logger` moved to `IdentityAccessApplication` 🚚 
- New class `Gatekeeper` in `IdentityAccessDomainModel` for access state logic 👶 
- New class `IdentityService` in `IdentityAccessDomainModel` for authentication actions 🛂 
- iOS App uses global `Authenticator` to keep sessionID and userID in memory after authentication 📝 

Boy, that's a lot of work.

### CODE REVIEW

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 1]
```diff
@@ -5,90 +5,187 @@
...
-            completion?(false)
-            return
+    public enum Error: Swift.Error, Hashable {
+        case emptyPassword
+    }
+
+    private var gatekeeperRepository: GatekeeperRepository { return DomainRegistry.gatekeeperRepository }
+    private var gatekeeper: Gatekeeper! { return gatekeeperRepository.gatekeeper() }
+    private var clock: Clock { return ApplicationServiceRegistry.clock }
+    private var identityService: IdentityService { return DomainRegistry.identityService }
```
Comment (DmitryBespalov) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: vars and methods unordered

File: IdentityAccess/IdentityAccessImplementations/ethereum/EncryptionService.swift [hunk 2]
```diff
@@ -50,4 +50,8 @@ public final class EncryptionService: EncryptionServiceProtocol {
         return Mnemonic(EthereumKit.Mnemonic.create())
     }
 
+    public func encrypted(_ plainText: String) -> String {
+        return plainText
```
Comment (DmitryBespalov) [inline IdentityAccess/IdentityAccessImplementations/ethereum/EncryptionService.swift:54]: implementation missing(next PRs)

File: IdentityAccess/IdentityAccessImplementations/persistence/InMemorySessionRepositoryTests.swift [hunk 3]
```diff
@@ -0,0 +1,19 @@
+//
+//  Copyright © 2018 Gnosis Ltd. All rights reserved.
+//
+
+import XCTest
+@testable import IdentityAccessImplementations
+import IdentityAccessDomainModel
+
+class InMemorySessionRepositoryTests: XCTestCase {
```
Comment (DmitryBespalov) [inline IdentityAccess/IdentityAccessImplementations/persistence/InMemorySessionRepositoryTests.swift]: rename and add tests

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 4]
```diff
@@ -5,90 +5,187 @@
...
 
-    open func authenticateUser(password: String? = nil, completion: ((Bool) -> Void)? = nil) {
-        if isAuthenticationBlocked {
-            completion?(false)
-            return
+    public enum Error: Swift.Error, Hashable {
+        case emptyPassword
+    }
+
+    private var gatekeeperRepository: GatekeeperRepository { return DomainRegistry.gatekeeperRepository }
```
Comment (sche) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: Need to discuss this part

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 5]
```diff
@@ -5,90 +5,187 @@
...
+    private var gatekeeper: Gatekeeper! { return gatekeeperRepository.gatekeeper() }
+    private var clock: Clock { return ApplicationServiceRegistry.clock }
+    private var identityService: IdentityService { return DomainRegistry.identityService }
+
+    private var isAccessPossible: Bool {
```
Comment (sche) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: isAccessable ?

File: Common/Common/IdentifiedEntity.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
+//
+//  Copyright © 2018 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+
+open class IdentifiedEntity<T: Hashable>: Hashable, Assertable {
```
Comment (sche) [inline Common/Common/IdentifiedEntity.swift]: IdentifiableEntity

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 7]
```diff
@@ -5,90 +5,187 @@
...
+
+    private init(_ method: AuthenticationMethod, _ password: String? = nil) {
+        precondition(method == .biometry && password == nil ||
+            method == .password && password != nil, "Invalid authentication request")
+        self.method = method
+        self.password = password
+    }
+
+    public static func biometry() -> AuthenticationRequest {
+        return AuthenticationRequest(.biometry)
```
Comment (sche) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: AuthenticationRequest(with: .biometry)

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 8]
```diff
@@ -5,90 +5,187 @@
...
+
+    public static let blocked = AuthenticationResult(status: .blocked, userID: nil, sessionID: nil)
+    public static let failure = AuthenticationResult(status: .failure, userID: nil, sessionID: nil)
+    public static func success(userID: String, sessionID: String) -> AuthenticationResult {
+        return AuthenticationResult(status: .success, userID: userID, sessionID: sessionID)
+    }
+}
 
-    private let account: AccountProtocol = Account.shared
+// this must be responsible for registration and authentication
```
Comment (sche) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: remove comment

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 9]
```diff
@@ -5,90 +5,187 @@
...
+    open func isAuthenticationMethodPossible(_ method: AuthenticationMethod) -> Bool {
+        guard isAccessPossible else { return false }
+        var possible: AuthenticationMethod = .password
+        if isAuthenticationMethodSupported(.faceID) && biometricService.isAuthenticationAvailable {
+            possible.insert(.faceID)
+        }
+        if isAuthenticationMethodSupported(.touchID) && biometricService.isAuthenticationAvailable {
+            possible.insert(.touchID)
+        }
+        return !method.isDisjoint(with: possible)
```
Comment (sche) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: not readable

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 10]
```diff
@@ -5,90 +5,187 @@
...
+    private var isAccessPossible: Bool {
+        return gatekeeper.isAccessPossible(at: clock.currentTime)
+    }
+
+    open func authenticateUser(_ request: AuthenticationRequest) throws -> AuthenticationResult {
+        let time = clock.currentTime
+        let user: UserDescriptor?
+        if request.method == .password {
+            user = try identityService.authenticateUser(password: request.password, at: time)
+        } else if request.method.isSubset(of: .biometry) {
```
Comment (sche) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: biometry.contains...

File: IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift [hunk 11]
```diff
@@ -5,90 +5,187 @@
...
-    open func configureMaxPasswordAttempts(_ count: Int) {
-        account.maxPasswordAttempts = count
+    open func configureBlockDuration(_ duration: TimeInterval) throws {
+        try gatekeeper.changeBlockDuration(duration)
+        try gatekeeperRepository.save(gatekeeper)
     }
 
-    open func configureBlockDuration(_ duration: TimeInterval) {
-        account.blockedPeriodDuration = duration
+    open func provisionAuthenticationPolicy(sessionDuration: TimeInterval,
```
Comment (sche) [inline IdentityAccess/IdentityAccessApplication/AuthenticationApplicationService.swift]: please rename

**Review by sche — APPROVED** [review]

**Pull request merged**